### PR TITLE
Falcon optimization

### PR DESCRIPTION
**Pull request by libinta**
# What does this PR do?
1. add use_flash_attentiong, flash_attention_recompute, flash_attention_causal_mask
2. add mark step per decoder
3. add fusedsdpa fp8

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ssarkar2** [comment]
measurement:
`QUANT_CONFIG=./quantization_config/maxabs_measure_include_outputs.json python run_generation.py --model_name_or_path mistralai/Mistral-7B-Instruct-v0.2 --attn_softmax_bf16 --use_hpu_graphs --trim_logits --use_kv_cache --reuse_cache --bf16 --batch_size 1 --warm 0 --n_iter 1 --flash_attention_recompute --flash_attention_causal_mask --use_flash_attention`

actual run:
`QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_generation.py --model_name_or_path tiiuae/falcon-40b  --use_hpu_graphs --use_kv_cache --limit_hpu_graphs --bucket_size 128 --max_new_tokens 2048 --batch_size 4 --bf16 --reuse_cache --bucket_internal --fp8 --max_input_tokens 2048 --warm 2 --n_iter 2 --flash_attention_recompute --flash_attention_causal_mask --use_flash_attention`


for 2048->2048, falcon 40b
 
with bs8:
without PR: 185 tps with 84.2 GB memory
with PR: 187 tps with 72.9 GB
 
with BS12:
without PR: OOM
with PR: 219 tps with 90.63 HB memory

File: examples/text-generation/README.md [hunk 1]
```diff
@@ -155,7 +155,8 @@ python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py \
 --use_hpu_graphs \
 --use_kv_cache \
 --batch_size 1 \
---do_sample
+--do_sample \
+--use_flash_attention --flash_attention_causal_mask
```
Comment (schoi-habana) [inline examples/text-generation/README.md:159]: please add these to run_lm_eval.py, too

Comment (libinta) [reply examples/text-generation/README.md:159]: @schoi-habana the fusedsdpa fp8 is not ready for falcon yet

File: optimum/habana/transformers/models/falcon/modeling_falcon.py [hunk 2]
```diff
@@ -241,13 +291,30 @@ class GaudiFalconAttention(FalconAttention):
...
-            self.sdpa = ScaledDotProductAttention(config)
+        """
+        Choice of SDPA:
+        There are these variables: use_flash_attention and datatype (bf16/fp8)
+        datatype is determined by presence of QUANT_CONFIG env var, presence of which indicates fp8
+        1. use_flash_attention, fp8: use ModuleFusedSDPA. most optimal
+        2. use_flash_attention, bf16: use FusedSDPA
+        3. not use_flash_attention, fp8: Use ScaledDotProductAttention, along with QUANT_CONFIG. This is the case before this PR
+        4. not use_flash_attention, bf16: F.scaled_dot_product_attention. Slowest option
+        """
```
Comment (regisss) [inline optimum/habana/transformers/models/falcon/modeling_falcon.py:310]: Let's move the docstring before the call to `super().__init__` right above

File: optimum/habana/transformers/models/falcon/modeling_falcon.py [hunk 3]
```diff
@@ -786,8 +896,12 @@ def forward(
         # head_mask has shape n_layer x batch x num_heads x N x N
         head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
 
-        htcore.mark_step()
+        # htcore.mark_step()
         for i, (block, layer_past) in enumerate(zip(self.h, past_key_values)):
+            # if not self.training and (
+            #    torch.distributed.is_initialized() is False or torch.distributed.get_world_size() == 1
+            # ):
+            # htcore.mark_step()
```
Comment (regisss) [inline optimum/habana/transformers/models/falcon/modeling_falcon.py:904]: Do we need to keep these commented lines?

**Review by regisss — APPROVED** [review]

**Review by regisss — APPROVED** [review]

**Pull request merged**